pkg/networking/backend_sg_provider.go

```diff
@@ -0,0 +1,245 @@
+	p.autoGeneratedSG = awssdk.StringValue(resp.GroupId)
+	return nil
+}
+
+func (p *defaultBackendSGProvider) buildBackendSGTags(_ context.Context) []*ec2sdk.TagSpecification {
+	return []*ec2sdk.TagSpecification{
+		{
+			ResourceType: awssdk.String(resourceTypeSecurityGroup),
+			Tags: []*ec2sdk.Tag{
+				{
+					Key:   awssdk.String(tagKeyK8sCluster),
+					Value: awssdk.String(p.clusterName),
+				},
+				{
+					Key:   awssdk.String(tagKeySGType),
+					Value: awssdk.String(tagValueBackend),
+				},
+			},
+		},
+	}
+}
+
+func (p *defaultBackendSGProvider) getBackendSGFromEC2(ctx context.Context, sgName string, vpcID string) (string, error) {
+	req := &ec2sdk.DescribeSecurityGroupsInput{
+		Filters: []*ec2sdk.Filter{
+			{
+				Name:   awssdk.String("vpc-id"),
+				Values: awssdk.StringSlice([]string{vpcID}),
+			},
+			{
+				Name:   awssdk.String("group-name"),
+				Values: awssdk.StringSlice([]string{sgName}),
+			},
+		},
+	}
+	p.logger.V(1).Info("Queriying existing SG", "vpc-id", vpcID, "name", sgName)
+	sgs, err := p.ec2Client.DescribeSecurityGroupsAsList(ctx, req)
+	if err != nil && !isEC2SecurityGroupNotFoundError(err) {
+		return "", err
+	}
+	if len(sgs) > 0 {
+		return awssdk.StringValue(sgs[0].GroupId), nil
+	}
+	return "", nil
+}
+
+func (p *defaultBackendSGProvider) isBackendSGRequired(ctx context.Context) (bool, error) {
+	ingList := &networking.IngressList{}
+	if err := p.k8sClient.List(ctx, ingList); err != nil {
+		p.logger.Error(err, "Unable to list ingresses")
+		return true, errors.Wrapf(err, "unable to list ingresses")
+	}
+	for _, ing := range ingList.Items {
+		for _, fin := range ing.GetFinalizers() {
+			if fin == implicitGroupFinalizer || strings.HasPrefix(fin, explicitGroupFinalizerPrefix) {
+				return true, nil
+			}
+		}
+	}
+	p.logger.Info("No ingress found, backend SG can be deleted", "SG ID", p.autoGeneratedSG)
+	return false, nil
+}
+
+func (p *defaultBackendSGProvider) releaseSG(ctx context.Context) error {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	if len(p.autoGeneratedSG) == 0 {
+		return nil
+	}
+
+	if required, err := p.isBackendSGRequired(ctx); required || err != nil {
+		p.logger.V(1).Info("backend SG is required, releaseSG ignore delete")
+		return err
+	}
+	req := &ec2sdk.DeleteSecurityGroupInput{
+		GroupId: awssdk.String(p.autoGeneratedSG),
+	}
+	p.logger.V(1).Info("deleting default backend SG", "ID", p.autoGeneratedSG)
+	if err := runtime.RetryImmediateOnError(2*time.Minute, 2*time.Second, isSecurityGroupDependencyViolationError, func() error {
+		_, err := p.ec2Client.DeleteSecurityGroupWithContext(ctx, req)
+		return err
+	}); err != nil {
+		return errors.Wrap(err, "failed to delete securityGroup")
+	}
+	p.logger.Info("deleted securityGroup", "ID", p.autoGeneratedSG)
+
+	p.autoGeneratedSG = ""
+	return nil
+}
+
+func (p *defaultBackendSGProvider) getBackendSGName() string {
+	sgNameHash := sha256.New()
+	_, _ = sgNameHash.Write([]byte(p.vpcID))
```

Comment:
since we have can same sgName across multiple vpcs, seems use vpcID as hash component is unnecessary here.